File: trace.py
#!/usr/bin/python
from __future__ import print_function
from bcc import BPF, USDT
from time import sleep

it, pid = open('pid', 'r').readline().split(',')
iteration = int(it)
prog = open('trace.c', 'r').read()
# hook #
prog = prog.replace('PID_FILTER_VOID', 'if(tgid != %s) return  ; ' % pid)
prog = prog.replace('PID_FILTER_ZERO', 'if(tgid != %s) return 0; ' % pid)
benchmark = USDT(pid=int(pid));

benchmark.enable_probe(probe="in_io_uring", fn_name="usdt_in_uring")
benchmark.enable_probe(probe="out_io_uring", fn_name="usdt_out_uring")
benchmark.enable_probe(probe="in_cq_advance", fn_name="usdt_in_advance")
benchmark.enable_probe(probe="out_cq_advance", fn_name="usdt_out_advance")

benchmark.enable_probe(probe="in_libaio", fn_name="usdt_in_native")
benchmark.enable_probe(probe="out_libaio", fn_name="usdt_out_native")

benchmark.enable_probe(probe="in_glibc_aio", fn_name="usdt_in_glibc")
benchmark.enable_probe(probe="out_glibc_aio", fn_name="usdt_out_glibc")

bpf = BPF(text=prog, usdt_contexts=[benchmark])
bpf.attach_uprobe(name="uring", sym="io_uring_submit", fn_name="trace_in_submit_uring")
bpf.attach_uretprobe(name="uring", sym="io_uring_submit", fn_name="trace_out_submit_uring")
bpf.attach_uprobe(name="uring", sym="io_uring_peek_batch_cqe", fn_name="trace_in_peek")
bpf.attach_uretprobe(name="uring", sym="io_uring_peek_batch_cqe", fn_name="trace_out_peek")

bpf.attach_uprobe(name="aio", sym="io_submit", fn_name="trace_in_submit_native")
bpf.attach_uretprobe(name="aio", sym="io_submit", fn_name="trace_out_submit_native")
bpf.attach_uprobe(name="aio", sym="io_getevents", fn_name="trace_in_poll_native")
bpf.attach_uretprobe(name="aio", sym="io_getevents", fn_name="trace_out_poll_native")


bpf.attach_uprobe(name="rt", sym="lio_listio", fn_name="trace_in_submit_glibc")
bpf.attach_uretprobe(name="rt", sym="lio_listio", fn_name="trace_out_submit_glibc")
bpf.attach_uprobe(name="rt", sym="aio_suspend", fn_name="trace_in_poll_glibc")
bpf.attach_uretprobe(name="rt", sym="aio_suspend", fn_name="trace_out_poll_glibc")

# listen #
try:
    sleep(10000)
except KeyboardInterrupt:
    print()

# data processsing (uring, native, posix) #

def avg_val(inp, outp):
    m, n = [list() for i in range(2)]
    for k,v in inp.items():
        m.append((k.value, v.value))
    for k,v in outp.items():
        n.append((k.value, v.value))
    avg = 0
    for (i,j) in zip(sorted(m),sorted(n)):
        avg += j[1]-i[1]
    return (avg / iteration)

avg_totals, avg_submits, avg_peeks  = [list() for i in range(3)]
est_maplist, call_maplist, peek_maplist, submit_maplist = [list() for i in range(4)]
est_maplist.extend([bpf["est_uring"], bpf["est_native"], bpf["est_glibc"]])
call_maplist.extend([bpf["sysc_uring"], bpf["sysc_native"], bpf["sysc_glibc"]])

peek_maplist.append((bpf["in_peek"], bpf["out_peek"]))
peek_maplist.append((bpf["in_poll_native"], bpf["out_poll_native"]))
peek_maplist.append((bpf["in_poll_glibc"], bpf["out_poll_glibc"]))

submit_maplist.append((bpf["in_submit_uring"], bpf["out_submit_uring"]))
submit_maplist.append((bpf["in_submit_native"], bpf["out_submit_native"]))
submit_maplist.append((bpf["in_submit_glibc"], bpf["out_submit_glibc"]))

for est_map in est_maplist:
    m = []
    avg_total = 0
    for k,v in est_map.items():
        m.append((k.value, v.value))
        if (v.value != 0):
            avg_total += v.value
    avg_totals.append(avg_total / iteration)

for (in_map, out_map) in submit_maplist:
    avg = avg_val(in_map, out_map)
    avg_submits.append(avg)

avg_advance = avg_val(bpf["in_advance"], bpf["out_advance"])

for (in_map, out_map) in peek_maplist:
    avg = avg_val(in_map, out_map)
    avg_peeks.append(avg)

for i in range(3):
    if (i == 0):
        print("io_uring")
        print("%20s %20s %20s %20s" %("total cost(us)", "submit cost", "poll cost", "clear cost"))
        print("%20d %20.4f %20.4f %20.4f" % (avg_totals[i]//1000, avg_submits[i]/avg_totals[i], avg_peeks[i]/avg_totals[i], avg_advance/avg_totals[i]))
        continue
    elif (i == 1):
        print("native aio")
    else:
        print("posix aio")
    print("%20s %20s %20s" %("total cost(us)", "submit cost", "poll cost"))
    print("%20d %20.4f %20.4f" % (avg_totals[i]//1000, avg_submits[i]/avg_totals[i], avg_peeks[i]/avg_totals[i]))

File: trace.c
#include <linux/sched.h>
/* TOTAL COST */
BPF_HASH(est_uring);
BPF_HASH(est_native);
BPF_HASH(est_glibc);

/* SUBMISSION COST */
BPF_HASH(in_submit_uring);
BPF_HASH(out_submit_uring);
BPF_HASH(in_submit_native);
BPF_HASH(out_submit_native);
BPF_HASH(in_submit_glibc);
BPF_HASH(out_submit_glibc);

/* POLL LOST*/
BPF_HASH(in_peek);
BPF_HASH(out_peek);
BPF_HASH(in_poll_native);
BPF_HASH(out_poll_native);
BPF_HASH(in_poll_glibc);
BPF_HASH(out_poll_glibc);
/* ADVANCE LOST */
BPF_HASH(in_advance);
BPF_HASH(out_advance);

/* SYSCALL COUNTER */
BPF_HASH(sysc_uring);
BPF_HASH(sysc_native);
BPF_HASH(sysc_glibc);

void usdt_in_uring (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u64 key, val;
    u64 *cnt; 
    
    /* update visit-times */
    key = 0;
    cnt = est_uring.lookup(&key);
    if (cnt == 0) {val = 1;}
    else {val = *cnt+1; }
    est_uring.update(&key, &val);

    /* update kernel time */
    est_uring.update(&val, &time); 
}

void usdt_out_uring (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u64 key, val;
    u64 *cnt; 
    
    /* lookup visit-times */
    key = 0;
    cnt = est_uring.lookup(&key);
    if (cnt == 0) {key = 1;}    // just for passing verifier
    else {key = *cnt; } 

    /* update delta time */    
    u64 *nsp = est_uring.lookup(&key);
    if (nsp != 0) {time -= *nsp;}
    else {time = 0;} 
    est_uring.update(&key, &time); 
}

void usdt_in_advance (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    //PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_uring.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = in_advance.lookup(&key);
    if (tsp != 0) time += *tsp;
    in_advance.update(&key, &time);
}

void usdt_out_advance (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    //PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_uring.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = out_advance.lookup(&key);
    if (tsp != 0) time += *tsp;
    out_advance.update(&key, &time);
}

void usdt_in_native (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u64 key, val;
    u64 *cnt; 
    
    /* update visit-times */
    key = 0;
    cnt = est_native.lookup(&key);
    if (cnt == 0) {val = 1;}
    else {val = *cnt+1; }
    est_native.update(&key, &val);

    /* update kernel time */
    est_native.update(&val, &time); 
}


void usdt_out_native (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u64 key, val;
    u64 *cnt; 
    
    /* lookup visit-times */
    key = 0;
    cnt = est_native.lookup(&key);
    if (cnt == 0) {key = 1;}    
    else {key = *cnt; } 

    /* update delta time */    
    u64 *nsp = est_native.lookup(&key);
    if (nsp != 0) {time -= *nsp;}
    else {time = 0;} 
    est_native.update(&key, &time); 
}

void usdt_in_glibc (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u64 key, val;
    u64 *cnt; 
    
    /* update visit-times */
    key = 0;
    cnt = est_glibc.lookup(&key);
    if (cnt == 0) {val = 1;}
    else {val = *cnt+1; }
    est_glibc.update(&key, &val);

    /* update kernel time */
    est_glibc.update(&val, &time); 
}

void usdt_out_glibc (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u64 key, val;
    u64 *cnt; 
    
    /* lookup visit-times */
    key = 0;
    cnt = est_glibc.lookup(&key);
    if (cnt == 0) {key = 1;}    
    else {key = *cnt; }     

    /* update delta time */    
    u64 *nsp = est_glibc.lookup(&key);
    if (nsp != 0) {time -= *nsp;}
    else {time = 0;} 
    est_glibc.update(&key, &time); 
}

/* one test may multi-submit    */
void trace_in_submit_uring (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_uring.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = in_submit_uring.lookup(&key);
    if (tsp != 0) time += *tsp;
    in_submit_uring.update(&key, &time);
}

void trace_out_submit_uring (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_uring.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = out_submit_uring.lookup(&key);
    if (tsp != 0) time += *tsp;
    out_submit_uring.update(&key, &time);
}

void trace_in_peek (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_uring.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = in_peek.lookup(&key);
    if (tsp != 0) time += *tsp;
    in_peek.update(&key, &time);
}

void trace_out_peek (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_uring.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = out_peek.lookup(&key);
    if (tsp != 0) time += *tsp;
    out_peek.update(&key, &time);
}

void trace_in_submit_native (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_native.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = in_submit_native.lookup(&key);
    if (tsp != 0) time += *tsp;
    in_submit_native.update(&key, &time);
}

void trace_out_submit_native (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_native.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = out_submit_native.lookup(&key);
    if (tsp != 0) time += *tsp;
    out_submit_native.update(&key, &time);
}

void trace_in_poll_native (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_native.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = in_poll_native.lookup(&key);
    if (tsp != 0) time += *tsp;
    in_poll_native.update(&key, &time);
}

void trace_out_poll_native (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_native.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = out_poll_native.lookup(&key);
    if (tsp != 0) time += *tsp;
    out_poll_native.update(&key, &time);
}

void trace_in_submit_glibc (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_glibc.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = in_submit_glibc.lookup(&key);
    if (tsp != 0) time += *tsp;
    in_submit_glibc.update(&key, &time);
}

void trace_out_submit_glibc (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_glibc.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = out_submit_glibc.lookup(&key);
    if (tsp != 0) time += *tsp;
    out_submit_glibc.update(&key, &time);
}

void trace_in_poll_glibc (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_glibc.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = in_poll_glibc.lookup(&key);
    if (tsp != 0) time += *tsp;
    in_poll_glibc.update(&key, &time);
}

void trace_out_poll_glibc (struct pt_regs *ctx) {
    u64 time = bpf_ktime_get_ns();
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_VOID
    u64 key = 0; u64 *visit, *tsp;

    /* get visit-times */
    visit = est_glibc.lookup(&key);
    if (visit == 0) return;

    /* add to total cost */
    key = *visit;
    tsp = out_poll_glibc.lookup(&key);
    if (tsp != 0) time += *tsp;
    out_poll_glibc.update(&key, &time);
}


TRACEPOINT_PROBE(syscalls, sys_enter_io_uring_setup) {
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_ZERO
    u64 *tmp;
    u64 key = 0, val = 1;
    tmp = sysc_uring.lookup(&key);
    if (tmp != 0) { val = *tmp+1; }
    sysc_uring.update(&key, &val);
    return 0;
}

TRACEPOINT_PROBE(syscalls, sys_enter_io_uring_enter) {
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_ZERO
    u64 *tmp;
    u64 key = 1, val = 1;
    tmp = sysc_uring.lookup(&key);
    if (tmp != 0) { val = *tmp+1; }
    sysc_uring.update(&key, &val);
    return 0;
}

TRACEPOINT_PROBE(syscalls, sys_enter_io_uring_register) {
    u32 tgid = bpf_get_current_pid_tgid() >> 32;
    PID_FILTER_ZERO
    u64 *tmp;
    u64 key = 2, val = 1;
    tmp = sysc_uring.lookup(&key);
    if (tmp != 0) { val = *tmp+1; }
    sysc_uring.update(&key, &val);
    return 0;
}
